File: support/src/main/java/com/wuguangxin/adapter/BaseRecyclerAdapter.java
package com.wuguangxin.adapter;

import android.content.Context;
import android.view.View;
import android.view.ViewGroup;

import com.wuguangxin.listener.OnItemClickListener;
import com.wuguangxin.utils.ToastUtils;

import java.util.ArrayList;
import java.util.List;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

public abstract class BaseRecyclerAdapter<T, V extends BaseViewHolder> extends RecyclerView.Adapter<V> {
    protected Context context;
    /** 数据列表 */
    private final List<T> mData = new ArrayList<>();
    /** 列表类型（可用于标识列表的类型） */
    private int listType;

    private OnItemClickListener<T> onItemClickListener;

    public BaseRecyclerAdapter(Context context) {
        this.context = context;
    }

    public BaseRecyclerAdapter(Context context, List<T> data) {
        this.context = context;
        this.mData.clear();
        if (data != null) {
            this.mData.addAll(data);
        }
    }

    @NonNull
    @Override
    final public V onCreateViewHolder(@NonNull ViewGroup viewGroup, int position) {
        View view = View.inflate(viewGroup.getContext(), getLayoutId(), null);
        view.setOnClickListener(v -> {
            if (onItemClickListener != null) {
                onItemClickListener.onItemClick(v, getItem(position), position, listType);
            }
        });
        return createViewHolder(view, position);
    }

    @Override
    public void onBindViewHolder(@NonNull V holder, int position) {
        bindViewData(holder, mData.get(position), position, getListType());
    }

    public abstract int getLayoutId();

    public V createViewHolder(View view, int position) {
        return (V) new BaseViewHolder(view);
    }

    public abstract void bindViewData(V holder, T data, int position, int type);

    @Override
    public int getItemCount() {
        return mData.size();
    }

    public Context getContext() {
        return context;
    }

    public List<T> getList() {
        return mData;
    }

    public void setListType(int listType) {
        this.listType = listType;
    }

    public int getListType() {
        return listType;
    }

    public T getItem(int position) {
        return mData.get(position);
    }

    public void setOnItemClickListener(OnItemClickListener<T> onItemClickListener) {
        this.onItemClickListener = onItemClickListener;
    }

    public OnItemClickListener<T> getOnItemClickListener() {
        return onItemClickListener;
    }

    public void setList(List<T> data) {
        this.mData.clear();
        if (data != null) {
            this.mData.addAll(data);
        }
        notifyAdapter();
    }

    public void notifyAdapter() {
        notifyDataSetChanged();
    }

    public void showToast(String text) {
        if (context == null) return;
        ToastUtils.showToast(context, text);
    }
}

File: support/src/main/java/com/wuguangxin/adapter/BaseRecyclerAdapter.kt
package com.wuguangxin.adapter

import android.content.Context
import android.util.Log
import android.view.View
import androidx.recyclerview.widget.RecyclerView
import android.view.ViewGroup
import com.wuguangxin.listener.OnItemClickListener
import com.wuguangxin.utils.ToastUtils

abstract class BaseRecyclerAdapter<T, V : BaseViewHolder?> : RecyclerView.Adapter<V> {
    private var context: Context?

    private val mData = mutableListOf<T>()

    var onItemClickListener: OnItemClickListener<T>? = null

    /** 列表类型（可用于标识列表的类型）  */
    var listType = 0

    constructor(context: Context) {
        this.context = context
    }

    constructor(context: Context?, data: List<T>?) {
        this.context = context
        mData.clear()
        if (data != null) {
            mData.addAll(data)
        }
    }

    override fun onCreateViewHolder(viewGroup: ViewGroup, position: Int): V {
        val view = View.inflate(viewGroup.context, getLayoutId(), null)
        view.setOnClickListener {
            Log.e("BaseRecyclerAdapter", "position = $position")
            onItemClickListener?.onItemClick(it, getItem(position), position, listType)
        }
        return createViewHolder(view, position)
    }

    override fun onBindViewHolder(holder: V, position: Int) {
        bindViewData(holder, mData[position], position, listType)
    }

    abstract fun getLayoutId(): Int

    open fun createViewHolder(view: View, position: Int): V {
        return BaseViewHolder(view) as V
    }

    abstract fun bindViewData(holder: V, data: T, position: Int, type: Int)

    override fun getItemCount(): Int {
        return mData.size
    }

    open var list: List<T>?
        get() = mData
        set(data) {
            mData.clear()
            data?.let {
                mData.addAll(it)
            }
            notifyDataSetChanged()
        }

    fun getItem(position: Int): T {
        return mData[position]
    }

    fun showToast(text: String?) {
        if (context == null) return
        ToastUtils.showToast(context, text)
    }
}